contracts/5/IbETHRouter.sol

```diff
@@ -0,0 +1,301 @@
+    function swapAlphaForExactETH(
+        uint256 amountETHOut,
+        uint256 amountAlphaInMax,         
+        address to,
+        uint256 deadline
+    ) external returns (uint256[] memory amounts) {
+        TransferHelper.safeTransferFrom(alpha, msg.sender, address(this), amountAlphaInMax);
+        address[] memory path = new address[](2);
+        path[0] = alpha;
+        path[1] = ibETH;
+        uint256[] memory swapAmounts = IUniswapV2Router02(router).swapTokensForExactTokens(exactETHToIbETH(amountETHOut), amountAlphaInMax, path, address(this), deadline);                           
+        IBank(ibETH).withdraw(swapAmounts[1]);
+        amounts = new uint256[](2);
+        amounts[0] = swapAmounts[0];
+        amounts[1] = address(this).balance;
+        TransferHelper.safeTransferETH(to, address(this).balance);        
+        if (amountAlphaInMax > amounts[0]) {
+            TransferHelper.safeTransfer(alpha, msg.sender, amountAlphaInMax.sub(amounts[0]));
+        }                    
+    }
+
+    // Swap exact amount of Token for ETH
+    // 1. Receive Token from caller
+    // 2. Swap Token for ibETH.
+    // 3. Unwrap ibETH to ETH.
+    function swapExactAlphaForETH(
+        uint256 amountAlphaIn,
+        uint256 amountETHOutMin,         
+        address to,
+        uint256 deadline
+    ) external returns (uint256[] memory amounts) {
+        TransferHelper.safeTransferFrom(alpha, msg.sender, address(this), amountAlphaIn); 
+        address[] memory path = new address[](2);
+        path[0] = alpha;
+        path[1] = ibETH;
+        uint256[] memory swapAmounts = IUniswapV2Router02(router).swapExactTokensForTokens(amountAlphaIn, 0, path, address(this), deadline);                        
+        IBank(ibETH).withdraw(swapAmounts[1]);        
+        amounts = new uint256[](2);
+        amounts[0] = swapAmounts[0];
+        amounts[1] = address(this).balance;
+        TransferHelper.safeTransferETH(to, amounts[1]);
+        require(amounts[1] >= amountETHOutMin, "IbETHRouter: receive less ETH than amountETHmin");                                       
+    }
+
+    // Swap ETH for exact amount of Token
+    // 1. Receive ETH from caller
+    // 2. Wrap ETH to ibETH.
+    // 3. Swap ibETH for Token    
+    function swapETHForExactAlpha(
+        uint256 amountAlphaOut,          
+        address to,
+        uint256 deadline
+    ) external payable returns (uint256[] memory amounts) {             
+        IBank(ibETH).deposit.value(msg.value)();              
+        uint256 amountIbETHInMax = IBank(ibETH).balanceOf(address(this));
+        address[] memory path = new address[](2);
+        path[0] = ibETH;
+        path[1] = alpha;
+        uint256[] memory swapAmounts = IUniswapV2Router02(router).swapTokensForExactTokens(amountAlphaOut, amountIbETHInMax, path, to, deadline);                
+        // Transefer left over ETH back
+        if (amountIbETHInMax > swapAmounts[1]) {            
```

Comment:
Should this be `swapAmounts[0]`? -> 0 for ibETH and 1 for ALPHA? Not sure if I am wrong or how the test passes.

Reply:
need to be swapAmounts[0]. The test pass because assertAlmostEqual tolerance is too loose to capture this case. I've fixed the test also to make sure that no dust ibETH is left in router.